ledger/controllers/grants.js

```diff
@@ -880,3 +1083,215 @@ module.exports.initialize = async (debug, runtime) => {
+    status: 'active',
+    promotionId
+  }
+  if (type === 'ads') {
+    Object.assign(query, {
+      type,
+      providerId: addresses.CARD_ID
+    })
+  }
+  return query
+}
+
+async function chooseFromAvailablePromotions (debug, runtime, {
+  entries
+}) {
+  let priority = Number.POSITIVE_INFINITY
+  const candidates = entries.reduce((memo, entry) => {
+    let candidates = memo
+    if (entry.priority > priority) {
+      return candidates
+    }
+    if (priority < entry.priority) {
+      candidates = []
+      priority = entry.priority
+    }
+    return candidates.concat(entry)
+  }, [])
+  if (!candidates.length) {
+    return candidates
+  }
+
+  const promotion = underscore.shuffle(candidates)[0]
+  return underscore.omit(promotion, unnecessaryPromotionDetails)
+}
+
+async function chooseFromAvailablePromotionsByType (debug, runtime, {
+  entries,
+  type
+}) {
+  const ads = underscore.filter(entries, ({
+    type
+  }) => type === 'ads')
+  const ugp = underscore.filter(entries, ({
+    type
+  }) => {
+    return type === '' || type == null || type === 'ugp'
+  })
+  const adsPromotions = await chooseFromAvailablePromotions(debug, runtime, {
+    entries: ads
+  })
+  const ugpPromotions = await chooseFromAvailablePromotions(debug, runtime, {
+    entries: ugp
+  })
+  return [].concat(adsPromotions, ugpPromotions)
+}
+
+function getGrantV4 (protocolVersion, createPayload) {
+  return (runtime) => async (request, reply) => {
+    const {
+      paymentId,
+      type
+    } = request.query
+    const query = {
+      active: true,
+      count: { $gt: 0 },
+      protocolVersion
+    }
+    const debug = braveHapi.debug(module, request)
+    const grants = runtime.database.get('grants', debug)
+    const promotions = runtime.database.get('promotions', debug)
+    const wallets = runtime.database.get('wallets', debug)
+
+    if (qaOnlyP(request)) {
+      return reply(boom.notFound())
+    }
+
+    if (paymentId) {
+      const promotionIds = []
+      const wallet = await wallets.findOne({ paymentId: paymentId })
+      if (!wallet) {
+        return reply(boom.notFound('no such wallet: ' + paymentId))
+      }
+      if (wallet.grants) {
+        wallet.grants.forEach((grant) => { promotionIds.push(grant.promotionId) })
+      }
+      underscore.extend(query, { promotionId: { $nin: promotionIds } })
+    }
+
+    const entries = await promotions.find(query)
+    if (!entries || !entries[0]) {
+      return reply(boom.notFound('no promotions available'))
+    }
+
+    const result = await createPayload(debug, runtime, {
+      entries,
+      type
+    })
+
+    try {
+      await Promise.all(result.map(async ({
+        promotionId
+      }) => {
+        const counted = await grants.count({ promotionId })
+        if (!counted) {
+          throw boom.notFound('promotion not available')
+        }
+      }))
+    } catch (e) {
+      return reply(e)
+    }
+
+    return reply({
+      grants: result
+    })
+  }
+}
```

Comment:
Would you help me understand the purpose of this endpoint? It looks like we want to return the number of grants available for a given promotion type, is that right?

Why do we want to extend the query like this if the wallet has grants?
```
underscore.extend(query, { promotionId: { $nin: promotionIds } })
```